### PR TITLE
Replaces ConnectorMetadata/BindingPath with the catalog interfaces.

### PR DESCRIPTION
**Pull request by RCHowell**
## Relevant Issues

#1496 

## Description

This PR replaces ConnectorMetadata with BindingPath/BindingName to use the Catalog interfaces such as Catalog, Identifier, Name, and Table. This is an intermediate step to replace the metadata layer with the catalog interfaces. This PR HAS to update the plugins, and it replaces PartiQLValue in the bindings to use Datum.

Finally, it introduces an initial implementation of the IonDatum which fills the role of the original Ion-valued backed ExprValue.

## Other Information
- Updated Unreleased Section in CHANGELOG: **NO**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by RCHowell** [comment]
This is a draft because I am working on confirming any impact to the conformance tests which @johnedquinn has worked so hard one 😁

File: partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                    Annotation.MISSING -> Datum.missing()
+                    Annotation.BAG -> Datum.nullValue(PType.typeBag())
+                    Annotation.DATE -> Datum.nullValue(PType.typeDate())
+                    Annotation.TIME -> Datum.nullValue(PType.typeTimeWithoutTZ(6))
+                    Annotation.TIMESTAMP -> Datum.nullValue(PType.typeTimeWithoutTZ(6))
+                    Annotation.GRAPH -> error("Datum does not support GRAPH type.")
+                    null -> Datum.nullValue()
+                }
+                BOOL -> when (tag) {
+                    null -> PType.typeBool()
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt:80]: Just a heads up -- you'll need to rebase from your most recently merged PR.

File: partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    override fun get(name: String): Datum {
+        if (_kind != STRUCT) {
+            return super.get(name)
+        }
+        // TODO handle multiple/ambiguous field names?
+        val v = _value.asStruct().getOptional(name)
+        // TODO handle nulls?
+        return if (v == null) {
+            Datum.nullValue()
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt:214]:
```suggestion
            Datum.missing()
```

File: partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        // TODO handle multiple/ambiguous field names?
+        val struct = _value.asStruct()
+        for (field in struct.fields) {
+            if (field.name.equals(name, ignoreCase = true)) {
+                return of(field.value)
+            }
+        }
+        // TODO handle missing fields?
+        return Datum.nullValue()
```
Comment (johnedquinn) [inline partiql-spi/src/main/kotlin/org/partiql/value/ion/IonDatum.kt:232]:
```suggestion
        return Datum.missing()
```

File: plugins/partiql-memory/README.md [hunk 4]
```diff
@@ -1,36 +1,19 @@
...
-
-The full path is `catalogName.schema.tbl`
-
-The lookup logic is identical to localPlugin. 
-
+// define the data and types.
+val catalog = MemoryCatalog.builder()
+    .name("hello")
+    .defineTable("pi", MemoryTable(
+        type = PType.typeFloat32(),
```
Comment (johnedquinn) [inline plugins/partiql-memory/README.md:12]:
```suggestion
        type = PType.real(),
```

File: plugins/partiql-local/src/main/kotlin/org/partiql/plugins/local/LocalCatalog.kt [hunk 5]
```diff
@@ -1,109 +1,109 @@
...
-internal class LocalCatalog private constructor(private val root: FsTree.D) {
+    override fun getName(): String {
+        return name
+    }
+
+    override fun getTable(session: Session, name: Name): Table? {
+        val path = toPath(name.getNamespace()).resolve(name.getName() + EXT)
+        if (path.notExists() || !path.isDirectory()) {
+            return null
+        }
```
Comment (johnedquinn) [inline plugins/partiql-local/src/main/kotlin/org/partiql/plugins/local/LocalCatalog.kt:39]:
```suggestion
        if (path.notExists() || path.isDirectory()) {
            return null
        }
```
If I'm reading this correctly.

File: plugins/partiql-local/src/main/kotlin/org/partiql/plugins/local/LocalCatalog.kt [hunk 6]
```diff
@@ -1,109 +1,109 @@
...
         }
+        return super.listTables(session, namespace)
     }
 
-    private suspend fun SequenceScope<BindingPath>.search(acc: List<BindingName>, node: FsTree.T) {
-        val steps = acc + BindingName(node.name, BindingCase.INSENSITIVE)
-        this.yield(BindingPath(steps))
+    override fun listNamespaces(session: Session, namespace: Namespace): Collection<Namespace> {
+        val path = toPath(namespace)
+        if (path.notExists() || path.isDirectory()) {
```
Comment (johnedquinn) [inline plugins/partiql-local/src/main/kotlin/org/partiql/plugins/local/LocalCatalog.kt:85]:
```suggestion
        if (path.notExists() || !path.isDirectory()) {
```

**Review by johnedquinn — CHANGES_REQUESTED** [review]

**Review by johnedquinn — APPROVED** [review]

**Pull request merged**